core/src/main/java/io/temporal/samples/hello/CommandProcessor.java

```diff
@@ -0,0 +1,137 @@
+package io.temporal.samples.hello;
+
+import io.temporal.activity.ActivityInterface;
+import io.temporal.activity.ActivityOptions;
+import io.temporal.api.enums.v1.WorkflowIdReusePolicy;
+import io.temporal.client.WorkflowClient;
+import io.temporal.client.WorkflowOptions;
+import io.temporal.client.WorkflowStub;
+import io.temporal.serviceclient.WorkflowServiceStubs;
+import io.temporal.worker.Worker;
+import io.temporal.worker.WorkerFactory;
+import io.temporal.workflow.CompletablePromise;
+import io.temporal.workflow.UpdateMethod;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.WorkflowInterface;
+import io.temporal.workflow.WorkflowMethod;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+
+public class CommandProcessor {
+  static final String TASK_QUEUE = "MyTaskQueue";
+  static final String WORKFLOW_ID = "MyWorkflowId";
+
+  @WorkflowInterface
+  public interface CommandProcessorWorkflow {
+    @WorkflowMethod
+    String startProcessing();
+
+    @UpdateMethod
+    String submitCommand(int command);
+  }
+
+  @ActivityInterface
+  public interface MyActivities {
+    String processCommand(int command);
+  }
+
+  public static class CommandProcessorWorkflowImpl implements CommandProcessorWorkflow {
+
+    private static class QueuedCommand {
+      public int command;
+      public CompletablePromise<String> promise;
+
+      public QueuedCommand(int command, CompletablePromise<String> promise) {
+        this.command = command;
+        this.promise = promise;
+      }
+    }
+
+    private ArrayList<QueuedCommand> commandQueue;
+    private boolean done;
+
+    public CommandProcessorWorkflowImpl() {
+      this.commandQueue = new ArrayList<>();
+      this.done = false;
+    }
+
+    private final MyActivities activities =
+        Workflow.newActivityStub(
+            MyActivities.class,
+            ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(10)).build());
+
+    @Override
+    public String startProcessing() {
+      while (true) {
+        Workflow.await(() -> this.commandQueue.size() > 0 || this.done);
+        if (this.done && this.commandQueue.size() == 0) {
+          return "done";
+        }
+        QueuedCommand queuedCommand = this.commandQueue.remove(0);
+        String result = activities.processCommand(queuedCommand.command);
+        queuedCommand.promise.complete(result);
+      }
+    }
+
+    @Override
+    public String submitCommand(int command) {
+      if (command < 0) {
+        this.done = true;
+        return "stopping workflow";
+      }
+      CompletablePromise<String> promise = Workflow.newPromise();
+      QueuedCommand queuedCommand = new QueuedCommand(command, promise);
+      this.commandQueue.add(queuedCommand);
+      return queuedCommand.promise.get();
```

Comment:
@Quinn-With-Two-Ns Is there a way here to free up the update thread? e.g. could the update handler return the promise, rather than block on `.get()`?

Reply:
Good point, I'd be interested to know this, but haven't looked into it yet.

Reply:
Yeah we can explore alternative approaches that use less threads. Given we only allow 10 concurrent updates it wasn't prioritized, virtual threads also make this not a big deal.